### PR TITLE
[Validator] Add AutoMapping constraint to enable or disable auto-validation

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #32070, #32015   
| License       | MIT
| Doc PR        | todo

As discussed in #32070 and #32015, it's sometimes mandatory to prevent some classes or properties to be auto mapped (auto-validated). This PR introduces a new constraint, `@AutoMapping` allowing to do exactly that. Examples:

Class:

```php
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * @ORM\Entity
 * @Assert\AutoMapping(false)
 */
class DoctrineLoaderNoAutoMappingEntity
{
    /**
     * @ORM\Id
     * @ORM\Column
     */
    public $id;

    /**
     * @ORM\Column(length=20, unique=true)
     */
    public $maxLength;
}
```

Property:

```php
namespace Symfony\Bridge\Doctrine\Tests\Fixtures;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * @ORM\Entity
 */
class DoctrineLoaderEntity extends DoctrineLoaderParentEntity
{
    /**
     * @ORM\Id
     * @ORM\Column
     */
    public $id;

    /**
     * @ORM\Column(length=10)
     * @Assert\AutoMapping(false)
     */
    public $noAutoMapping;
}
```

The rules are the following:

* If the constraint is present on a property, and set to true, auto-mapping is always on, regardless of the config, and of any class level annotation
* If the constraint is present on a property, and set to false, auto-mapping is always off, regardless of the config, and of any class level annotation
* If the constraint is present on a class, and set to true, auto-mapping is always on except if a the annotation has been added to a specific property, and regardless of the config
* If the constraint is present on a class, and set to false, auto-mapping is always off except if a the annotation has been added to a specific property, and regardless of the config

### CODE REVIEW

**Comment by noniagriconomie** [comment]
Hi @dunglas, this means in sf 4.3 it is a default feature (doctrine field are alwas by default validated) ?
the dev need to optout to get rid off it? if yes, why not the other side?
thank you

**Comment by nicolas-grekas** [comment]
@noniagriconomie see https://github.com/symfony/recipes/pull/612

BTW, I think we also need an opt-in annotation
having both config enabling+exclusion list and annotations to enable/disable looks the most sensible to me. Every style covered, maximum flexibility.

**Comment by OskarStark** [comment]
I agree with Nicolas here, Opt-In is definitely needed too 👍🏻

**Comment by noniagriconomie** [comment]
thx for both answers @nicolas-grekas @OskarStark :)

File: src/Symfony/Component/Validator/CHANGELOG.md
Comment (alanpoulain) [inline src/Symfony/Component/Validator/CHANGELOG.md]:
```suggestion
* added `NoAutoMapping` constraint to opt-out a class or a property from auto mapping
```

File: src/Symfony/Component/Validator/Constraints/NoAutoMapping.php
Comment (alanpoulain) [inline src/Symfony/Component/Validator/Constraints/NoAutoMapping.php]: To remove.

File: src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php
Comment (alanpoulain) [inline src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php]:
```suggestion
    private function supports(ClassMetadata $metadata, string $classValidatorRegexp = null): bool
```

**Comment by alanpoulain** [comment]
IMHO it should be enabled by default in the recipes (and allow to opt-out some classes or properties) and having an opt-in feature is a nice to have.

**Comment by fabpot** [comment]
As we experienced issues when we enabled it by default and side effects that are difficult to understand, I'm in favor of only supporting the opt-in mode and only for classes/properties.

**Comment by dunglas** [comment]
Being opt-in only defects the purpose of this feature and decreases the security benefits it provides: https://github.com/symfony/symfony/issues/32070#issuecomment-502867215

Most side effects are because this feature is young, this PR should be enough to prevent the last remaining one.

The feature is already opt-in using the config file (it’s off by default). I don’t understand how the AutoMapping constraint would work? I mean, you must already opt in a specific directory using the config, and you can exclude some classes or properties using this new annotation. How a the proposed annotation would integrate with this?

File: src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php
Comment (dunglas) [inline src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: This is a bug fix

**Comment by nicolas-grekas** [comment]
Opt-in can still be made the default in maker-bundle, so that the security you want @dunglas is still a thing.

**Comment by dunglas** [comment]
Many people (including me) don't use maker bundle to create entities (for instance, almost all PHPStorm users, I guess). And it's one more annotation to not forget to add to every new class, for no reasons. Actually, I prefer the current situation (having to uncomment one time for all some lines in a config file) than having to add an annotation to all new entities.

**Comment by dunglas** [comment]
Travis failures are expected (we'll have to bump the required Validator version or to add conflict rules when merged).

**Comment by nicolas-grekas** [comment]
That's why I think we need config + annotation, opt-in/out + list/exclusion, to cover all styles.

**Comment by dunglas** [comment]
Here is a new proposal that should satisfy everybody:

* Change the constraint to `@AutoMapping(true/false)`, use `true` is the default value.
* If the constraint is present on a property, and set to `true`, auto-mapping is always on, regardless of the config, and of a class level level annotation
* If the constraint is present on a property, and set to `false`, auto-mapping is always off, regardless of the config, and of a class level level annotation
* If the constraint is present on a class, and set to `true`, auto-mapping is always on except if a the annotation has been added to a specific property, and regardless of the config
* If the constraint is present on a class, and set to `false`, auto-mapping is always off except if a the annotation has been added to a specific property, and regardless of the config
* The config stays unchanged (with the new behavior, there is no need to a config for exclusions, it's already handled by the constraint) ; and is uncommented by default in the recipe

It should cover all use cases, the feature is still on by default for the users of the framework thanks to the recipe. Do you folks agree?

**Comment by nicolas-grekas** [comment]
Annotations always win, looks sensible to me.
I would still like to have the exclusion list in the config, but someone else can provide it in another PR.

**Comment by nicolas-grekas** [comment]
Should target 4.4 to me, except the bug fix part of course.

**Comment by weaverryan** [comment]
With this proposal, if no annotations are added (assuming we re-add the auto mapping config in the recipe in 4.4), then auto mapping would still be turned on, correct? We could then, in make:entity decide to always generate entities with AutoMapping(true) to be explicit and make it more obvious how to turn it off (or even answer the question “where is this validation coming from?”). I think that sensible... though we may debate later whether to have the default config be “always on without annotation” or “off, unless the annotation is present” (or an I misunderstanding how the config would work?)

Thanks for the conversation. This feature makes me a bit uncomfortable... but @dunglas may very well be correct that it’s just this that’s needed to make it shine. After all, the form fields themselves hav built-in, automatic “sanity” validation, and that’s never been a problem (not an exact comparison).

**Comment by spackmat** [comment]
I think disabling all automatic constraints at once is a bit rough. Think, for example, of a decimal property type-hinted to float (or a Decimal object or whatever) instead of Doctrine's internal type string. At the moment, this leads to an invalid `Type` constraints. Disabling all automatic constraints for that property would mean to give up the wanted ones. This wouldn't be a big problem as the missing ones can easily be readded explicitly and as it is an Annotation, a dev will see that the automatic is disabled. But doing that the correct way requires devs to know about them all and keep track, when new ones will be added in the future.

Wouldn't it be less invasive to only disable the specifically problematic automatic constraints? Like `@Assert\NoAutoMapping(['Type', 'NotNull'])`?

**Comment by dunglas** [comment]
@spackmat https://github.com/symfony/symfony/pull/32107 should allow to deal with this problem

**Comment by dunglas** [comment]
I implemented this: https://github.com/symfony/symfony/pull/32107#issuecomment-504125587

Excluding only a specific annotation is not supported yet, but it would be nice to add this option (in a subsequent PR).

**Comment by ihmels** [comment]
Should this feature really be implemented with a validation constraint? In my opinion it feels wrong to enable or disable the auto-mapping by adding an constraint.

**Comment by dunglas** [comment]
It's similar to `@Valid` and `@Traverse`, it doesn't look wrong to me, and allows to reuse all the existing infrastructure (formats, cache...).

File: src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: the 2nd args to preg_quote are not needed, let's remove them to reduce the diff

File: src/Symfony/Component/Validator/Constraints/AutoMapping.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/AutoMapping.php]: I think we can make this text clearer by making it shorter. e.g.
> Using the annotations on a property has higher precedence than using it on a class, which has higher precedence than any configuration that might be defined outside the class.

File: src/Symfony/Component/Validator/Constraints/AutoMapping.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/AutoMapping.php]: missing final dot and double quotes around `%s`

File: src/Symfony/Component/Validator/Tests/Mapping/Loader/PropertyInfoLoaderTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Mapping/Loader/PropertyInfoLoaderTest.php]: 2nd args should be removed

File: src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php]: `isAutoMappingEnabledForClass` would be more clear

**Review by nicolas-grekas — APPROVED** [review]
LGTM, calling for another vote after minor comments are addressed.

File: src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php
Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: shouldn't it be fixed on 4.3 or lower?

File: src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php
Comment (xabbuh) [inline src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: I don't agree with this change. It causes a regression (see my comment at https://github.com/symfony/symfony/pull/33828#pullrequestreview-297993708).

File: src/Symfony/Component/Validator/Constraints/AutoMapping.php
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/AutoMapping.php]: What about having two dedicated annotations (e.g. `EnableAutoMapping` and `DisableAutoMapping`) instead of the option switch?

File: src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php
Comment (xabbuh) [inline src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php]: Should we mark the trait as internal?

**Review by xabbuh — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php
Comment (dunglas) [reply src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: If removed, PHPStorm (or PHP Inspections, I'm not sure) display a big error...

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: We don't care, tools don't have a brain :)

File: src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php
Comment (dunglas) [reply src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: And anyway it must not be in this PR. Reverting here, let's discuss on #33828

File: src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php
Comment (dunglas) [reply src/Symfony/Component/Validator/Mapping/Loader/AutoMappingTrait.php]: I don't think so, it helps to create custom loaders, and to prevent inconsistency such as the one introduced by  #33828

File: src/Symfony/Component/Validator/Constraints/AutoMapping.php
Comment (dunglas) [reply src/Symfony/Component/Validator/Constraints/AutoMapping.php]: Your call!

File: src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php
Comment (dunglas) [reply src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: Just reverting it isn't possible actually. We need to decide what behavior we want.

Comment (dunglas) [reply src/Symfony/Bridge/Doctrine/Tests/Validator/DoctrineLoaderTest.php]: Should be resolved in #33828 now.

**Comment by dunglas** [comment]
This one should be ready to be merged now. The last decision to make is between this single annotation and `EnableAutoMapping`+`DisableAutoMapping` as suggested by @xabbuh. Ping @symfony/mergers

**Comment by nicolas-grekas** [comment]
I think I prefer Enable/DisableAutoMapping because it opens for better extensibility.
One idea I have in mind would be to handle an `auto_mapper` attribute to the `validator.auto_mapper` tag:

`<tag name="validator.auto_mapper" auto_mapper="property_info" />`

Then `@Assert\DisableAutoMapping("property_info")` would allow specifically disabling this auto mapper (or enable it).

**Comment by dunglas** [comment]
`EnableAutoMapping` and `DisableAutoMapping` added. It should be ready to be merged now. /cc @xabbuh

**Review by nicolas-grekas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Comment by xabbuh** [comment]
Thank you Kévin.

**Pull request merged**
